### PR TITLE
TransformerEngine fallback compute dtype

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Per title

----
:books: Documentation preview :books:: https://pytorch-lightning--19082.org.readthedocs.build/en/19082/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/plugins/precision/transformer_engine.py [hunk 1]
```diff
@@ -42,13 +42,14 @@ class TransformerEnginePrecision(Precision):
...
     Args:
-        dtype: The weights dtype to use.
+        weights_dtype: The weights dtype to use.
         recipe: Recipe for the DelayedScaling
             `configuration <https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/common.html#transformer_engine.common.recipe.DelayedScaling>`__.
             In dict format or the dataclass format.
         replace_layers: Whether to replace ``Linear`` and ``LayerNorm`` layers automatically with their Transformer
             Engine alternatives. Note that they don't subclass the torch equivalents so checks like
             ``isinstance(l, torch.nn.Linear)`` will not pass.
+        fallback_compute_dtype: The compute dtype to use for operations that don't support fp8 autocast.
```
Comment (awaelchli) [inline src/lightning/fabric/plugins/precision/transformer_engine.py:52]:
```suggestion
        fallback_compute_dtype: The compute dtype to use for operations that don't support fp8 autocast. Defaults to the same as ``weights_dtype``.
```

File: src/lightning/fabric/CHANGELOG.md [hunk 2]
```diff
@@ -15,11 +15,17 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 
+- Added `TransformerEnginePrecision(fallback_compute_dtype=)` to control the dtype of operations that don't support fp8 ([#19082](https://github.com/Lightning-AI/lightning/pull/19082))
+
+
 ### Changed
 
 - `seed_everything()` without passing in a seed no longer randomly selects a seed, and now defaults to `0` ([#18846](https://github.com/Lightning-AI/lightning/pull/18846))
 
 
+- Changed the `TransformerEnginePrecision(dtype=)` argument to `weights_dtype` ([#19082](https://github.com/Lightning-AI/lightning/pull/19082))
```
Comment (awaelchli) [inline src/lightning/fabric/CHANGELOG.md:26]:
```suggestion
- Changed the `TransformerEnginePrecision(dtype=)` argument to `weights_dtype` and made it required ([#19082](https://github.com/Lightning-AI/lightning/pull/19082))
```

File: src/lightning/fabric/plugins/precision/transformer_engine.py [hunk 3]
```diff
@@ -80,21 +83,27 @@ def __init__(
...
-        self.dtype = dtype
+        self.weights_dtype = weights_dtype
         self.recipe = recipe
         self.replace_layers = replace_layers
+        self.fallback_compute_dtype = fallback_compute_dtype or weights_dtype
 
     def convert_module(self, module: torch.nn.Module) -> torch.nn.Module:
         # avoid converting if any is found. assume the user took care of it
-        if self.replace_layers and not any("transformer_engine.pytorch" in m.__module__ for m in module.modules()):
+        if any("transformer_engine.pytorch" in m.__module__ for m in module.modules()):
```
Comment (awaelchli) [inline src/lightning/fabric/plugins/precision/transformer_engine.py:93]: why remove the `self.replace_layers`? This now would raise the warning regardless, every time you do init-module.

File: src/lightning/pytorch/CHANGELOG.md [hunk 4]
```diff
@@ -32,6 +35,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - The `LightningModule.load_from_checkpoint()` function now calls `.configure_model()` on the model if it is overridden, to ensure all layers can be loaded from the checkpoint ([#19036](https://github.com/Lightning-AI/lightning/pull/19036))
 
 
+- Changed the `TransformerEnginePrecision(dtype=)` argument to `weights_dtype` ([#19082](https://github.com/Lightning-AI/lightning/pull/19082))
```
Comment (awaelchli) [inline src/lightning/pytorch/CHANGELOG.md:38]:
```suggestion
- Changed the `TransformerEnginePrecision(dtype=)` argument to `weights_dtype` and made it required ([#19082](https://github.com/Lightning-AI/lightning/pull/19082))
```

File: src/lightning/fabric/plugins/precision/transformer_engine.py (hunk 3 above)
Comment (carmocca) [reply src/lightning/fabric/plugins/precision/transformer_engine.py:93]: When you do `Fabric(precision="transformer-engine")` then replace_layers is `None`. You don't get the message.

As you pointed out, you'd get it with `init_module` and `TransformerEnginePrecision(replace_layers=True)`. The comment below says this. It's a info message not a warning.

Do you have another suggestion?

**Review by awaelchli — APPROVED** [review]

File: src/lightning/fabric/plugins/precision/transformer_engine.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/fabric/plugins/precision/transformer_engine.py:93]: Ok, there is another `if replace_layers` nested under this if-statement, I didn't see that. I feel like there is a more intuitive way to simplify these if-else statements, right now it is hard to understand. But I'm too lazy to find a better way 😅

**Review by Borda — APPROVED** [review]

**Pull request merged**